Fix S3242: Should not suggest base class for arguments when method has an event handler signature

### Description

Hi there,

If you create a blank new UWP App you will get a S3242 false positive. The analyser suggests to use a `INavigationFailedEventArgs` instead of `NavigationFailedEventArgs` in the event handler for the `Frame.OnNavigationFailed` event handler. But `INavigationFailedEventArgs` is inaccsessible due to its protection level.

### Repro steps
I created a small [repo solution]((https://github.com/SonarSource/sonar-csharp/files/1233687/S3242FalsePositive.zip) with an UWP app. In the App.xaml.cs you can see the warning in row 71. If you change the parameter type of `args` to `INavigationFailedEventArgs` you will get a compiler error CS0122

### Expected behavior

No S3242 warning on an UWP event handler if the genreal type of a parameter is inaccsessible due to its protection level.

### Actual behavior

A S3242 warning on an UWP event handler if the genreal type of a parameter is inaccsessible due to its protection level.

### Known workarounds

No workaround known.

### Related information

* SonarC# Version: 6.3.0.2862
* Visual Studio Version: 15.3.0

--- a/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs
+++ b/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs
@@ -60,7 +60,8 @@ namespace SonarAnalyzer.Rules.CSharp
             if (methodSymbol == null ||
                 methodSymbol.Parameters.Length == 0 ||
                 methodSymbol.IsOverride ||
-                methodSymbol.GetInterfaceMember() != null)
+                methodSymbol.GetInterfaceMember() != null ||
+                IsEventHandlerImplementation(methodSymbol))
             {
                 return Enumerable.Empty<Diagnostic>();
             }
@@ -234,6 +235,17 @@ namespace SonarAnalyzer.Rules.CSharp
             return overridenSymbol != null
                 ? overridenSymbol.ContainingType
                 : accessedMember.ContainingType;
+        }
+
+        private bool IsEventHandlerImplementation(IMethodSymbol methodSymbol)
+        {
+            // Cannot find a way to precisely tell whether this method is a event handler implementation
+            // without doing too much browsing in all trees.
+            // Let's go for a simple heuristic.
+            return methodSymbol.ReturnsVoid &&
+                methodSymbol.Parameters.Length == 2 &&
+                methodSymbol.Parameters[0].Name == "sender" &&
+                methodSymbol.Parameters[0].Type.Is(KnownType.System_Object);
         }
 
         private class ParameterData

--- a/sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodsShouldUseBaseTypes.cs
+++ b/sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodsShouldUseBaseTypes.cs
@@ -624,3 +624,31 @@ namespace Test_22
         }
     }
 }
+
+// Test that rule doesn't suggest other types for EventHandler methods
+namespace Test_24
+{
+    public interface IFooEventArgs
+    {
+        bool IsFoo { get; }
+    }
+    public class FooEventArgs : EventArgs, IFooEventArgs
+    {
+        public bool IsFoo { get; set; }
+    }
+
+    public class Foo
+    {
+        public event EventHandler<FooEventArgs> FooEvent;
+
+        public Foo()
+        {
+            FooEvent += Foo_FooEvent;
+        }
+
+        private void Foo_FooEvent(object sender, FooEventArgs e)
+        {
+            var x = e.IsFoo;
+        }
+    }
+}

